src/api/trace/span_context.rs

```diff
@@ -111,6 +114,110 @@ impl SpanId {
     }
 }
 
+/// TraceState carries system-specific configuration data, represented as a list
+/// of key-value pairs. TraceState allows multiple tracing systems to
+/// participate in the same trace.
+///
+/// Please review the [W3C specification] for details on this field.
+///
+/// [W3C specification]: https://www.w3.org/TR/trace-context/#tracestate-header
+#[cfg_attr(feature = "serialize", derive(Deserialize, Serialize))]
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct TraceState {
+    data: HashMap<String, String>,
+    ordered_keys: VecDeque<String>,
+}
+
+impl TraceState {
+    /// Creates a new `TraceState` from the given key-value collection.
+    pub fn from_key_value(trace_state: Vec<KeyValue>) -> Self {
+        let mut data: HashMap<String, String> = HashMap::with_capacity(trace_state.len());
+        let mut ordered_keys: VecDeque<String> = VecDeque::with_capacity(trace_state.len());
+
+        for kv in trace_state {
+            data.insert(kv.key.clone().into(), kv.value.into());
+            ordered_keys.push_back(kv.key.into());
+        }
+
+        TraceState { data, ordered_keys }
+    }
+
+    /// Retrieves a value for a given key from the `TraceState` if it exists.
+    pub fn get(&self, key: String) -> Option<String> {
```

Comment:
This could probably return `Option<&str>`, let the caller `clone` if they want owned data